File: 287.find-the-duplicate-number.go
/*
 * @lc app=leetcode id=287 lang=golang
 *
 * [287] Find the Duplicate Number
 */

package main

// @lc code=start
func findDuplicate(nums []int) int {
	fast := nums[0]
	slow := nums[0]
	for {
		slow = nums[slow]
		fast = nums[nums[fast]]
		if slow == fast {
			break
		}
	}

	slow2 := nums[0]
	for slow != slow2 {
		slow = nums[slow]
		slow2 = nums[slow2]
	}
	return slow
}

// func findDuplicate(nums []int) int {
// 	array := make([]bool, len(nums))
// 	for _, num := range nums {
// 		if array[num] {
// 			return num
// 		} else {
// 			array[num] = true
// 		}
// 	}
// 	return -1
// }

// @lc code=end

File: 287.find-the-duplicate-number.ts
/*
 * @lc app=leetcode id=287 lang=typescript
 *
 * [287] Find the Duplicate Number
 */

// @lc code=start
// function findDuplicate(nums: number[]): number {
//   for (let i = 0; i < nums.length; i++) {
//     for (let j = 0; j < nums.length; j++) {
//       if (i === j) {
//         continue;
//       }
//       if (nums[i] === nums[j]) {
//         return nums[i];
//       }
//     }
//   }
// }
function findDuplicate(nums: number[]): number {
  let listCopy = new Array(nums.length);
  for (let n of nums)
    if (listCopy[n] == undefined) {
      listCopy[n] = true;
    } else {
      return n;
    }
  return -1;
}
// function findDuplicate(nums: number[]): number {
//   let listCopy = new Array(10_001);
//   for (let n of nums)
//     if (listCopy[n] == undefined) {
//       listCopy[n] = true;
//     } else {
//       return n;
//     }
//   return -1;
// }
// @lc code=end
